packages/react-native/ReactCommon/react/renderer/textlayoutmanager/platform/ios/react/renderer/textlayoutmanager/RCTAttributedTextUtils.mm

```diff
@@ -144,7 +144,10 @@ inline static CGFloat RCTEffectiveFontSizeMultiplierFromTextAttributes(const Tex
           isnan(textAttributes.fontSize) ? RCTBaseSizeForDynamicTypeRamp(dynamicTypeRamp) : textAttributes.fontSize;
       return [fontMetrics scaledValueForValue:requestedSize] / requestedSize;
     } else {
-      return textAttributes.fontSizeMultiplier;
+      Float fontSizeMultiplier = textAttributes.fontSizeMultiplier;
+      Float maxFontSizeMultiplier =
+          isnan(textAttributes.maxFontSizeMultiplier) ? 0.0 : textAttributes.maxFontSizeMultiplier;
+      return maxFontSizeMultiplier >= 1.0 ? fminf(maxFontSizeMultiplier, fontSizeMultiplier) : fontSizeMultiplier;
```

Comment:
@Abbondanzo was suggesting that this implementation could lead to unexpected bugs because `maxFontSizeMultiplier` is not used in the branch above:
```c++
  if (textAttributes.allowFontScaling.value_or(true)) {
    if (textAttributes.dynamicTypeRamp.has_value()) {
```

His suggestion is to refactor the code to something like (roughly, please **double check this!**):

```c++
inline static CGFloat RCTEffectiveFontSizeMultiplierFromTextAttributes(const TextAttributes &textAttributes)
{
  if (textAttributes.allowFontScaling.value_or(true)) {
    CGFloat fontSizeMultiplier = !isnan(textAttributes.fontSizeMultiplier) ? textAttributes.fontSizeMultiplier : 1.0;
    if (textAttributes.dynamicTypeRamp.has_value()) {
      DynamicTypeRamp dynamicTypeRamp = textAttributes.dynamicTypeRamp.value();
      UIFontMetrics *fontMetrics =
          [UIFontMetrics metricsForTextStyle:RCTUIFontTextStyleForDynamicTypeRamp(dynamicTypeRamp)];
      CGFloat requestedSize =
          isnan(textAttributes.fontSize) ? RCTBaseSizeForDynamicTypeRamp(dynamicTypeRamp) : textAttributes.fontSize;  
      fontSizeMultiplier = [fontMetrics scaledValueForValue:requestedSize] / requestedSize;
    }
    CGFloat maxFontSizeMultiplier = !isnan(textAttributes.maxFontSizeMultiplier) ? textAttributes.maxFontSizeMultiplier : 0.0;
    return maxFontSizeMultiplier >= 1.0 ? fminf(maxFontSizeMultiplier, fontSizeMultiplier) : fontSizeMultiplier;
  } else {
    return 1.0;
  }
}
```

WDYT? Could you apply the change and test if it works properly?

Reply:
He'd be correct. That's a mistake from me, comparing with the old architecture implementation. I'm admittedly a bit confused about expected behaviour of Dynamic Type on iOS as I've never used the feature, or at least the RN Tester example confuse me by using `fontSize` and `dynamicTypeRamp`. Need to understand the expected behaviour there before I can be confident my PR is doing what it should. I'll figure it out over the weekend!

Reply:
Good catch @Abbondanzo !

I didn't understand that Dynamic Type was scaling non-linearly at first, I thought it was just essentially just setting a predefined font size, so I didn't understand why we were specifying font size in the examples, but it makes sense now.

Here's before and after applying the suggested fix. It should align with the old architecture. Added some more RN Tester examples for dynamic type ramp while testing as well.

| Before  | After |
| ------------- | ------------- |
| ![Simulator Screenshot - iPhone 15 Pro - 2024-12-07 at 10 53 21](https://github.com/user-attachments/assets/ed583a9a-67af-4c16-be57-cc9bb0d62e12) | ![Simulator Screenshot - iPhone 15 Pro - 2024-12-07 at 10 52 24](https://github.com/user-attachments/assets/0c66fcd3-9470-4faf-9c0a-3e256b0f7387) |